catalog/src/main/java/org/killbill/billing/catalog/api/user/DefaultCatalogUserApi.java

```diff
@@ -89,9 +89,9 @@ public DefaultCatalogUserApi(final CatalogService catalogService,
             return fullCatalog;
         } else {
             final DefaultVersionedCatalog filteredCatalog = new DefaultVersionedCatalog(clock);
-            for (final StandaloneCatalog v : fullCatalog.getVersions()) {
+            for (final StaticCatalog v : fullCatalog.getVersions()) {
                 if (v.getEffectiveDate().compareTo(catalogDateVersion.toDate()) >= 0) {
-                    filteredCatalog.add(v);
+                    filteredCatalog.add((StandaloneCatalog) v);
```

Comment:
Does `DefaultVersionedCatalog` need to maintain `StandaloneCatalog` versions just because of JAXB (instead of `StaticCatalog`)? If so, would it be easier to hide these casts inside `DefaultVersionedCatalog` so they are not propagated around in the code?

Reply:
I see... you really want me to fix these ugly casts 🤣 

I started to make the changes, and use the `@XmlElement#type()` method to keep the interface and still allow JAXB to work properly; this leads to removing some casts but also adds a whole new set. Mostly our objects export quite a few internal apis (`initialize()`, `validate()`,  optimized versions of our `find` apis, ... This was adding more casts than removing so i undid my changes, but i am ready to have the discussion with you again...